### ISSUE TITLE
retrieve_tokens() takes long

### ISSUE THREAD

**Issue by StefanoDePascale** [QLVL/nephosem#10]
The retrieve_tokens() function takes a much longer time than in previous versions of the code, and the memory usage is also much higher than previous versions.

**Comment by montesmariana** [comment]
The issue seems to be when the results are converted to a dictionary of `TypeNode` objects (which was similar in the previous implementation, so I don't know why it would be slower here) and probably also in the following step of turning it into a matrix.
The bar in the picture above is completed when all the tokens are brought together into one dictionary `type2toks` where the keys are the types (one in each call, in this case) and the values are lists of tokens.
Until then, we don't have that much of an issue.

Then we have:

```python
type2toks = {k: TypeNode.merge(v) for k, v in type2toks.items()}
```
This line of code is very similar to its counterpart in the previous version. The following step is turning the dictionary into a matrix. I added a log message before and after this merging function with a timer to check how long it takes, so I can see whether that is the bottleneck or it's the creation of the matrix instead (or both).

![image](https://user-images.githubusercontent.com/54317904/166822114-92e75b60-9a84-4a9d-9647-3c8cc6def466.png)

The image shows that, for the type with 12k tokens:
- the full recollection took 1.15 minutes
- of which 12 seconds were spent distributing files across processors
- and 5 seconds to merging their output into one dictionary.
- The final function to turn the dictionary values into merged `TypeNode` objects took 3.4 seconds,
- and then the matrix was created in 4.28 seconds.

With the much larger (131x) type, instead, I haven't finished the large recollection;
- distributing the tasks took 13 seconds (this depends on the corpus size, not the number of tokens to collect),
- merging them into one dictionary started much later (it wasn't recorded, but it took a few minutes) and took 4 minutes with 18 seconds (only 51 times more than the other sample)
- and turning the values into `TypeNode` objects took 11 minutes and 41 seconds (about 163 times more than the smaller sample).

After about three hours, there is no sight of the matrix yet.
In other words, turning the dictionary into a dictionary of TypeNodes and then creating a matrix have a complexity such that very big groups become too demanding.
I'm not sure how to accelerate that, but in the meantime, my recommendation is to split it into smaller groups, by creating different matrices for each lemma/concept or scanning fewer files at a time.
